sentry_streams/sentry_streams/adapters/arroyo/translator.py

```diff
@@ -0,0 +1,246 @@
+)
+
+TPayload = TypeVar("TPayload")
+TResult = TypeVar("TResult")
+
+logger = logging.getLogger(__name__)
+
+
+class ArroyoAccumulator:
+
+    def __init__(
+        self,
+        acc: Callable[[], Accumulator[Any, Any]],
+    ) -> None:
+        self.acc = acc
+
+    def initial_value(self) -> Any:
+        # instantaiate the underlying accumulator every time
+        self.instance = self.acc()
+
+        # get the fresh initial value each time
+        return self.instance.get_value()
+
+    # types still tbd
+    def accumulator(self, result: Any, value: BaseValue[RoutedValue]) -> RoutedValue:
+        self.instance.add(value.payload.payload)
+
+        routed = RoutedValue(
+            route=value.payload.route,
+            payload=self.instance.get_value(),
+        )
+
+        return routed
+
+
+class WindowedReduce(
+    ProcessingStrategy[Union[FilteredPayload, TPayload]], Generic[TPayload, TResult]
+):
+    def __init__(
+        self,
+        window_size: Union[int, float],
+        window_slide: Union[int, float],
+        acc: Callable[[], Accumulator[Any, Any]],
+        next_step: ProcessingStrategy[TResult],
+    ) -> None:
+
+        window_count = math.ceil(window_size / window_slide)
+
+        self.window_count = window_count
+        self.window_size = window_size
+        self.window_slide = window_slide
+        self.acc = acc
+        self.next_step = next_step
+
+        # initializers depending on whether it is count-based or time-based
+        if isinstance(window_size, int):
+            self.count_mode = True
+            self.message_count = 1
+
+        else:
+            self.count_mode = False
+            self.start_time = time.time()
+
+        # build the list of Reduces, as we know the number of windows to manage upfront
+        self.windows = [self.create_reduce() for _ in range(self.window_count)]
+
+        # we know the starting value of every window goes through a cycle
+        # this represents the size of that cycle / loop
+        self.window_loop = self.window_count * self.window_slide
+
+        # since we're adding elements to a cycle effectively,
+        # we know, for each window, which indices within that cycle
+        # CANNOT be part of that window
+
+        # for each window, track a 'greater' and a 'lesser' value
+        # every incoming message will be compared to these to determine
+        # whether or not the message can be added to the window
+        self.boundaries = []
+        for i in range(self.window_count):
+            greater = (self.window_size + i * self.window_slide) % self.window_loop
+
+            # this branch is because of the different meaning of window_size
+            # between count and time
+            # e.g. count window of 5 msgs: [msg1 ... msg5]
+            # e.g. time window of 5 seconds [0s ... 5s]
+            if self.count_mode:
+                less = i * self.window_slide + 1
+            else:
+                less = i * self.window_slide
```

Comment:
For sliding window I think we should stick to time based windows only.
I think I may have said the other way around before but I realized this now. 
The sliding window is mostly useful with constant increments, which means time based window. 
It should simplify your implementation